File: backend/app/api/v1/endpoints/firms.py
from uuid import UUID

from fastapi import (
    APIRouter,
    Depends,
    Response,
    status,
)
from sqlalchemy.orm import Session

from app.dependencies.auth import (
    get_current_user,
)
from app.dependencies.database import get_db
from app.models.user import User
from app.schemas.firm import (
    FirmCreate,
    FirmResponse,
    FirmUpdate,
)
from app.services.firm_service import (
    FirmService,
)

router = APIRouter()


@router.post(
    "",
    response_model=FirmResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_firm(
    request: FirmCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = FirmService(db)

    return service.create_firm(
        request,
        current_user,
    )


@router.get(
    "",
    response_model=list[FirmResponse],
)
def list_firms(
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = FirmService(db)

    return service.list_firms(
        current_user
    )


@router.get(
    "/{firm_id}",
    response_model=FirmResponse,
)
def get_firm(
    firm_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = FirmService(db)

    return service.get_firm(
        firm_id,
        current_user,
    )


@router.put(
    "/{firm_id}",
    response_model=FirmResponse,
)
def update_firm(
    firm_id: UUID,
    request: FirmUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = FirmService(db)

    return service.update_firm(
        firm_id,
        request,
        current_user,
    )


@router.delete(
    "/{firm_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_firm(
    firm_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(
        get_current_user
    ),
):

    service = FirmService(db)

    service.delete_firm(
        firm_id,
        current_user,
    )

    return Response(
        status_code=status.HTTP_204_NO_CONTENT
    )

File: frontend/src/api/firms.ts
import api from "./client";

/* ===========================
   Models
=========================== */

export interface Firm {
  id: string;
  name: string;
  gst_number: string | null;
  phone: string | null;
  email: string | null;
  address: string | null;
  is_active: boolean;
}

export interface FirmRequest {
  name: string;
  gst_number?: string;
  phone?: string;
  email?: string;
  address?: string;
  is_active?: boolean;
}

/* ===========================
   APIs
=========================== */

export async function getFirms(): Promise<Firm[]> {
  const response = await api.get<Firm[]>("/firms");
  return response.data;
}

export async function getFirm(
  id: string,
): Promise<Firm> {
  const response = await api.get<Firm>(
    `/firms/${id}`,
  );

  return response.data;
}

export async function createFirm(
  request: FirmRequest,
): Promise<Firm> {
  const response = await api.post<Firm>(
    "/firms",
    request,
  );

  return response.data;
}

export async function updateFirm(
  id: string,
  request: FirmRequest,
): Promise<Firm> {
  const response = await api.put<Firm>(
    `/firms/${id}`,
    request,
  );

  return response.data;
}

export async function deleteFirm(
  id: string,
): Promise<void> {
  await api.delete(`/firms/${id}`);
}
